File: src/main/java/uk/laxd/dndSimulator/action/Turn.java
package uk.laxd.dndSimulator.action;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import uk.laxd.dndSimulator.character.Character;
import uk.laxd.dndSimulator.event.EncounterEventFactory;
import uk.laxd.dndSimulator.event.EventLogger;

public class Turn {

    private static final Logger LOGGER = LoggerFactory.getLogger(Turn.class);

    private final EncounterEventFactory eventFactory;
    private final EventLogger eventLogger;
    private final ActionResolver actionResolver;
    private final TargetSelector targetSelector;

    private Character character;

    public Turn(EncounterEventFactory eventFactory, EventLogger eventLogger, ActionResolver actionResolver, Character character, TargetSelector targetSelector) {
        this.eventFactory = eventFactory;
        this.eventLogger = eventLogger;
        this.actionResolver = actionResolver;
        this.character = character;
        this.targetSelector = targetSelector;
    }

    public void doTurn() {
        eventLogger.logEvent(eventFactory.createTurnStartEvent(this));

        Character target = targetSelector.getTarget();

        if(target == null) {
            LOGGER.info("Nothing to attack");
            return;
        }

        MeleeAttackAction attackAction = new MeleeAttackAction(character, character.getWeapon(), target);

        actionResolver.resolve(attackAction);

        LOGGER.debug(attackAction.toString());
    }

    public Character getCharacter() {
        return character;
    }
}

File: app/src/main/kotlin/uk/laxd/dndSimulator/action/Turn.kt
package uk.laxd.dndSimulator.action

import org.slf4j.LoggerFactory
import uk.laxd.dndSimulator.event.EncounterEventFactory
import uk.laxd.dndSimulator.character.Character
import uk.laxd.dndSimulator.event.EventLogger
import uk.laxd.dndSimulator.event.TurnEndEvent
import uk.laxd.dndSimulator.event.TurnStartEvent

class Turn(
    private val actionFactory: ActionFactory,
    private val actionResolver: ActionResolver,
    val character: Character,
    private val targetSelector: TargetSelector
) {
    private val logger = LoggerFactory.getLogger(Turn::class.java)

    fun doTurn() {
        EventLogger.instance.logEvent(TurnStartEvent(character))

        repeat(character.attackCount) {
            logger.debug("Attack $it/${character.attackCount}")

            val attackAction = actionFactory.createAction(character, targetSelector)
            actionResolver.resolve(attackAction)
            logger.debug(attackAction.toString())
        }

        EventLogger.instance.logEvent(TurnEndEvent(character))
    }
}